ext/msgpack/packer.h

```diff
@@ -357,6 +357,13 @@ static inline void msgpack_packer_write_map_header(msgpack_packer_t* pk, unsigne
 
 static inline void msgpack_packer_write_ext(msgpack_packer_t* pk, int ext_type, VALUE payload)
 {
+    if (payload == Qnil) {
+        msgpack_buffer_ensure_writable(PACKER_BUFFER_(pk), 3);
+        msgpack_buffer_write_2(PACKER_BUFFER_(pk), 0xc7, 0);
+        msgpack_buffer_write_1(PACKER_BUFFER_(pk), ext_type);
+        return;
+    }
```

Comment:
I can't understand why this is added. Could you explain why this is needed?

Reply:
It's for this call: https://github.com/msgpack/msgpack-ruby/pull/261/files/727edafda4e2b44e53a2b800dea7b286735c1865#diff-c177757220580d2500819916eee890a84548208c724d85c67a5eb41bde0f8906R135

We pass `Qnil` as a payload to signal that we're in "manual" serialization. When it happens we write a single byte extension that contains a NULL byte.

Reply:
Oh, it means, this new feature writes the size `0` for this ext type object, right?
Because it cannot determine its size before calling the Proc.

Wait. It should break other msgpack deserializer implementations...

Reply:
> Oh, it means, this new feature writes the size 0 for this ext type object, right?

Yes, exactly.

> Wait. It should break other msgpack deserializer implementations...

Well, my reasoning was that if the other implementation don't know about this extension type, they can't deserialize it anyway (unless they have some kind of basic handling of unknown extension types).

Reply:
The [spec of type `fext 8`](https://github.com/msgpack/msgpack/blob/master/spec.md#ext-format-family) says:
* ext 8 stores an integer and a byte array whose length is upto (2^8)-1 bytes
* XXXXXXXX is a 8-bit unsigned integer which represents N
* N is a length of data

If an implementation sets `0` into N, other implementations will stop reading bytes at the end of 3rd bytes, then it will try to deserialize the next objects - and all following objects will not be deserialized. It should be helped.

The reason for the heading length for ext types is, unpackers will be able to skip reading those bytes if it doesn't know what the ext type is. And the following bytes (other objects) will be deserialized after skipping it. It is the conversion that all packers/unpackers should follow.

I'm sorry that I missed this point at the time when I foud my original idea. I'll try to find workarounds or any other ideas.

Reply:
No worries. I thought it was ok. 

Well, let me know if you find another idea. 

Otherwise we might have to buffer what the extension writes to be able to set the proper size, which might be tricky.

Reply:
My very rough idealist (have never concerned any feasibilities):
* initialize a new buffer (B) in the packer, call the Proc with the packer with B, then restore the original buffer (A), write the length and type-id, concatenate B into A
* create a recorder object (new class, X) which behaves like a Packer, call the Proc with X, then the packer calculates the length to be appended and replay the recorded write operations on the buffer

(The `Recorder` object is like an object, which has methods like `write_nil`, `write_int`, ... and just stores the list of methods and arguments called. It can calculate the byte size to be appended, and also call the series of methods with arguments on the Packer.)

Reply:
> initialize a new buffer (B) in the packer

That sounds doable.

Another idea would be to make sure the buffer isn't flushed, and then find a way to go back and adjust the size, no idea if it's doable though.

Reply:
After a quick code check, I found [this line](https://github.com/msgpack/msgpack-ruby/blob/master/ext/msgpack/buffer_class.c#L194). My current preferable idea is:

1. initialize a new Buffer (B) in the packer
2. replace internal buffer (A) with B
3. call the Proc
4. restore A in the packer
5. write the length of B, ext type id
6. append the content of B

We may be able to optimize the Buffer-Buffer append operation as commented there (I don't have solid idea for now).

Reply:
> Another idea would be to make sure the buffer isn't flushed, and then find a way to go back and adjust the size, no idea if it's doable though.

This requires a completely different usage of Buffer, so I'm not positive about this.

Reply:
> My current preferable idea is:

Makes sense. I'll try to find some time to implement this.

Reply:
Ok, I did a quick hack just to see how it would work. The code is disgusting, and probably has a bunch of bugs, but it seems to work for simple cases.

That will definitely need a sizeable refactoring to be clean though :/

Reply:
I roughly checked the lines of `packer.c` L142-161 and it looks good to me.
Under this concept, these if-clause for `payload == nil` should be removed, right?

Reply:
> Under this concept, these if-clause for `payload == nil` should be removed, right?

Yes.